lcls_tools/common/measurements/profile.py

```diff
@@ -0,0 +1,169 @@
+        if self.fitting_tool is not None:
+            beam_statistics = self.fitting_tool.fit_images(images)
+            for name, val in beam_statistics:
+                outputs[name] = val
+
+        # save data if requested, add filename to outputs
+        if self.save_data:
+            fname = self._generate_new_filename()
+            self._write_data_to_hdf5(
+                np.array(images), fname, outputs)
+            outputs["save_filename"] = outputs
+
+        return outputs, images
+
+    def test_measurement(self):
+        """test the beam profile measurement"""
+        outputs, images = self.measure_profile(1, use_conditions=False)
+
+        print(outputs)
+        fig, ax = plt.subplots()
+        c = ax.imshow(images[0], origin="lower")
+        fig.colorbar(c)
+
+        plt.show()
+
+    def acquire_images(self, n_shots: int = 1, threaded=False, use_conditions=True):
+        """
+        Collect and saves images to HDF5.
+        Option for threading which spawns a child process
+        in a way that dependant GUIs/Code do not hang.
+        The extra_metadata dictionary will be attached
+        to the metadata for each image in HDF5 .
+
+        """
+        if threaded:
+            work = Thread(
+                target=self._take_images,
+                args=[n_shots, use_conditions],
+            )
+            # normally we join after start, but that blocked the pyqt main thread
+            # so we do not join here. If it breaks, look here first..
+            work.start()
+        else:
+            return self._take_images(n_shots, use_conditions)
+
+    def _take_images(self, n_shots: int = 1, use_conditions: bool = True):
+        """
+        Performs the work for collecting images and associated scalar data for a set
+        number of shots. Conditions such as charge windowing are applied.
+
+        """
+
+        images = []
+        scalar_measurements = []
+        while len(images) < n_shots:
+            time.sleep(self.wait_time_in_seconds)
+
+            # get image and PV's at the same time
+            img = self.device.get_processed_image()
+
+            # get extra pv values
+            result = {ele.pvname: ele.caget() for ele in self.extra_pvs}
+
+            # check to make sure all measurement conditions are satisfied
+            # if a condition is not satisfied then continue
+            if use_conditions:
+                for condition in self.condition_list:
+                    is_satisfied = condition.evaluate()
+                    result = result | condition.measurements()
+                    if not is_satisfied:
+                        continue
+
+            # if all conditions are satisfied then add the image and extra pv data to
+            # the list
+            images += [img]
+            scalar_measurements += [result]
+
+        return images, scalar_measurements
+
+    def _generate_new_filename(self, extension: Optional[str] = ".h5") -> str:
```

Comment:
I always wondered whether this type of stuff should be brought into it's own submodule i.e. `lcls_tools.common.io` or something. This would allow every file to have the same iso timestamp functionality. It would probably involve passing in a file-name for the user, and the extension is already taken care of.